modules/apps/site/site-cms-site-initializer/src/main/java/com/liferay/site/cms/site/initializer/internal/model/listener/ObjectEntryFolderModelListener.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2025 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.site.cms.site.initializer.internal.model.listener;
+
+import com.liferay.object.model.ObjectEntryFolder;
+import com.liferay.portal.kernel.exception.ModelListenerException;
+import com.liferay.portal.kernel.feature.flag.FeatureFlagManagerUtil;
+import com.liferay.portal.kernel.model.BaseModelListener;
+import com.liferay.portal.kernel.model.ModelListener;
+import com.liferay.portal.kernel.model.ResourceConstants;
+import com.liferay.portal.kernel.model.Role;
+import com.liferay.portal.kernel.security.permission.ActionKeys;
+import com.liferay.portal.kernel.service.ResourcePermissionLocalService;
+import com.liferay.site.cms.site.initializer.internal.util.CMSRoleUtil;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Jürgen Kappler
+ */
+@Component(service = ModelListener.class)
+public class ObjectEntryFolderModelListener
+	extends BaseModelListener<ObjectEntryFolder> {
+
+	@Override
+	public void onAfterCreate(ObjectEntryFolder objectEntryFolder)
+		throws ModelListenerException {
+
+		if (!FeatureFlagManagerUtil.isEnabled(
+				objectEntryFolder.getCompanyId(), "LPD-17564")) {
+
+			return;
+		}
+
+		try {
+			Role role = CMSRoleUtil.addAdminRoleAndPermissions(
+				objectEntryFolder.getCompanyId());
```

Comment:
Why do we have to call this every time we create a folder? Wouldn't it be better to do it only once when creating an instance, and then make this listener assume that the role already exists?